### ISSUE TITLE
Multiple tours with multiple steps

### ISSUE THREAD

**Issue by s8132** [benmarch/angular-ui-tour#68]
Hi,
I try create multiple tours with multiple steps and I have problem. I created two tours and first tour is ok, but when I start second tour I can't go to second step in second tour (error in the picture). 

``` html
<body ui-tour="main" ui-tour-backdrop="true" ui-tour-on-ready="readyTour(tour)">
<aside tour-step
               tour-step-title="Title1"
               tour-step-content="Content1"
               tour-step-order="0"
               tour-step-orphan="true"
               tour-step-backdrop="true"
               tour-step-animation="true"
               tour-step-append-to-body="true"
               tour-step-placement="top"
               tour-step-popup-delay="1"
               tour-step-next-path="main.howItWorks"
               tour-step-template-url="'views/tour/template.html'">
        </aside>
<aside tour-step
               tour-step-title="Title1"
               tour-step-content="Content1"
               tour-step-order="1"
               tour-step-orphan="true"
               tour-step-backdrop="true"
               tour-step-animation="true"
               tour-step-append-to-body="true"
               tour-step-placement="top"
               tour-step-popup-delay="1"
               tour-step-next-path="main.howItWorks"
               tour-step-template-url="'views/tour/template.html'">
        </aside>
<aside  tour-step="df" tour-step-belongs-to="reviewTour"
            tour-step-title="Title2"
            tour-step-content="Next tour1"
            tour-step-order="0"
            tour-step-orphan="true"></aside>

    <aside  tour-step="fd" tour-step-belongs-to="reviewTour"
            tour-step-title="Title2"
            tour-step-content="Next tour2"
            tour-step-order="1"
            tour-step-orphan="true"></aside>
```

``` js
uiTourService.createDetachedTour('reviewTour', {backdrop: true, useUiRouter: true, templateUrl: 'views/tour/template.html'});
    $rootScope.goToReviewTour = function(){
        // DemoState.goToDemo();
        uiTourService.getTourByName('main').end();
        uiTourService.getTourByName('reviewTour').start();
    };
    $rootScope.readyTour = function(tour){
        $rootScope.$on('$stateChangeSuccess', function () {
            tour.start();


            // if(!$cookies.get("tour")){
            //     a.start();
            // }
        });


    };
```

![zrzut ekranu 2016-08-10 o 21 59 01](https://cloud.githubusercontent.com/assets/1121745/17568830/c21f7c4c-5f45-11e6-8b46-2c5cb0ac7e4f.png)

In "reviewTour" tour show only first step and buttons not working (Next, End tour...).

**Comment by benmarch** [comment]
Hey @s8132, sorry for the delay. What versions of AngularJS, Angular Bootstrap, and Angular UI Tour are you using? I will attempt to recreate the issue locally.

**Comment by s8132** [comment]
Versions:
- angular: 1.5.7
- angular-bootstrap: 1.1.2
- angular-ui-tour: 0.5.5

I found second bug. If I set tour-step-template-url option then sometimes step has wrong position.

**Comment by benmarch** [comment]
Hey @s8132, I used your code to add a detached tour to my demo site and called the goToReviewTour() function and it appeared to work correctly for me. Are your tour steps loaded into the DOM and compiled by Angular prior to starting the tour? Try doing the following:

```js
uiTourService.getTourByName('main').end().then(function () { 
    uiTourService.getTourByName('reviewTour').start();
});
```

For the template issue, try preloading the template into the $templateCache. If that works then there might be a bug related to template loading. If that does not work, make sure all the required classes are in your provided template. You can find them in https://github.com/benmarch/angular-ui-tour/blob/0.5.5/app/templates/tour-step-template.html

Let me know how that goes.

**Comment by skalamichal** [comment]
Hello, 

I have similar issue (maybe the same) - I try to use multiple tours to teach the user to work with our SPA. I'd like to display different tour for each view, so I created multiple detached tours. When I run the detached tour with:
```
uiTourService.getTour().end().then(function() {
   uiTourService.getTourByName('mainTour').start();
});
```
I get following error:
```
angular.js:12808 TypeError: Cannot read property 'config' of null
    at self.goTo (angular-ui-tour.js:745)
    at self.next (angular-ui-tour.js:705)
    at fn (eval at compile (angular.js:13653), <anonymous>:4:260)
    at expensiveCheckFn (angular.js:14635)
    at callback (angular.js:24103)
    at Scope.$eval (angular.js:16384)
    at Scope.$apply (angular.js:16484)
    at HTMLButtonElement.<anonymous> (angular.js:24108)
    at defaultHandlerWrapper (angular.js:3384)
    at HTMLButtonElement.eventHandler (angular.js:3372)
```
I was able to track down the issue - one element visibility is controlled via the ng-if directive. And when the tour goes to this element (the step3 in the example tour), it's not possible to close the tour or continue to the next/previous tour step. See the HTML cope:
```
<ui-tour ui-tour-backdrop="true" ui-tour-placement="auto bottom" ui-tour-append-to-body="true">

	<div class="row">
		<div class="col-xs-12">
			<div class="form-group">
				<button class="btn btn-primary" ng-click="ctrl.start()" type="button">
					Start Tour
				</button>
				<button class="btn btn-default" ng-click="ctrl.visible = !ctrl.visible" type="button">
					Toggle visibility
				</button>
			</div>
		</div>
	</div>

	<div class="row">
		<div class="col-xs-6">
			<div class="panel panel-default"
				 tour-step="step1"
				 tour-step-title="Welcome!"
				 tour-step-content="Some Text 1"
				 tour-step-order="1"
				 tour-step-belongs-to="mainTour"
				 tour-step-next-step="step2">
				<div class="panel-heading">
					1st step
				</div>
				<div class="panel-body">
					This is panel for 1st step.
				</div>
			</div>
		</div>
		<div class="col-xs-6">
			<div class="panel panel-danger"
				 tour-step="step2"
				 tour-step-title="Welcome!"
				 tour-step-content="Some Text 2"
				 tour-step-order="2"
				 tour-step-next-step="step3"
				 tour-step-belongs-to="mainTour">
				<div class="panel-heading">
					2nd step
				</div>
				<div class="panel-body">
					This is panel for 2nd step.
				</div>
			</div>
		</div>
	</div>

	<div class="row" ng-if="ctrl.visible"
		 tour-step="step3"
		 tour-step-title="Doesn't work"
		 tour-step-content="Won't remove."
		 tour-step-order="3"
		 tour-step-belongs-to="mainTour"
		 tour-step-next-step="step4">
		<div class="col-xs-6 col-xs-offset-3">
			<div class="panel panel-warning">
				<div class="panel-heading">
					3rd step
				</div>
				<div class="panel-body">
					Tour stops here.
				</div>
			</div>
		</div>
	</div>

	<div class="row">
		<div class="col-xs-8 col-xs-offset-2">
			<div class="panel panel-info"
				 tour-step="step4"
				 tour-step-title="Welcome!"
				 tour-step-content="Some Text 4"
				 tour-step-order="4"
				 tour-step-belongs-to="mainTour">
				<div class="panel-heading">
					4th step
				</div>
				<div class="panel-body">
					This is panel for 4th step.
				</div>
			</div>
		</div>
	</div>

</ui-tour>
```
Is there a way to solve this issue? Any help would be helpful.

Thanks,
Michal

**Comment by benmarch** [comment]
Hey Michal,

Is the error occuring immediately when the tour starts, or when navigating to step 3?

If it is just when navigating, you might need to wrap the entire `.row` in an empty div and place the tourStep on that outside of the ngIf. Internally, when a tourStep is removed from the DOM (via ngIf) it also removes it from the tour, but if it is the middle of processing the step when it is removed then it will throw the NPE.

If it happens right away I will have to look into it more.

You can also remove the `tour-step-next-step` attributes since those are just for navigational purposes, and are used alongside `tour-step-next-path`.

**Comment by skalamichal** [comment]
Hi Ben,

thanks for quick response. It happens when I try to navigate from the step 3, to the next or previous step. The element is visible when the tour starts, just the visibility is controlled by via the ngIf directive. It's a bit strange, because when I list all steps from the tour (using the private _getSteps method), all steps are in the list.
I'll try to wrap the element in a new div.

Thanks, 
Michal

**Comment by skalamichal** [comment]
Hi Ben,

wrapping the element in a new div helped. 

Thanks,
Michal

**Comment by skalamichal** [comment]
Hi Ben,

the issue above exists in more areas, when I'm using the detached tour:
- can not use the tour-step within ng-if (I was able to fix this by wrapping the element in a new div, but it's not possible in all situations)
- can not use the tour-step within ng-include (I could solve this by including the template directly in to the page, but I don't think it's the best solution, again, it would fix this only in some situations)
- can not use the directive in ng-view - I teach the user to use navigation in the page (in the main index.html page) and then I'd like to continue with elements in the displayed view (within the ng-view inbox.html view) - the tour stops to work

If I don't use the detached tour, everything above works. The bug is always the same:
`angular.js:12808 TypeError: Cannot read property 'config' of null`

Is there any limitation with the detached tour?

Thanks,
Michal

**Comment by skalamichal** [comment]
I may have the fix, works for me. Code in the tourStep directive needs to be changed from:
```
//Add step to tour
scope.tourStep = step;
scope.tour = scope.tour || ctrl;
```
to
```
//Add step to tour
scope.tourStep = step;
scope.tour = ctrl;
```
The ngIf, ngInclude, ... directive create new scope. When element is removed and compiled back the scope.tour is not referenced correctly to the detachedTour, but to the main tour. This way we always have reference to the detached tour. And it seems nothing else needs to be updated.

Michal

**Comment by benmarch** [comment]
Ah good find!! Thank you, Michal! Would you like to create a PR, or would you rather I make the change?

**Comment by skalamichal** [comment]
Hi Ben,

I can do the PR if you prefer it - but it's really small change, so what do you like best.

Michal

**Comment by benmarch** [comment]
Hey Michal,

I would like to give you credit for it, so if you create the PR then your commit will always be in the repo history :)

Ben

**Comment by skalamichal** [comment]
OK, will do it ;)

**Comment by benmarch** [comment]
Closing this issue as resolved by #91.
